test/alphabet/alphabet_test.cpp

```diff
@@ -0,0 +1,337 @@
+// ==========================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ==========================================================================
+//
+// Copyright (c) 2006-2017, Knut Reinert, FU Berlin
+// Copyright (c) 2016-2017, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+
+#include <gtest/gtest.h>
+#include <sstream>
+
+#include <seqan3/alphabet.hpp>
+
+using namespace seqan3;
+
+template <typename T>
+class alphabet : public ::testing::Test
+{};
+
+// add all alphabets here
+using alphabet_types = ::testing::Types<dna4, dna5, rna4, rna5, nucl16,
+                                        /*aa27,*/
+                                        /*gap, gapped<nucl16>, */
+                                        illumina18, dna4q>;
+
+TYPED_TEST_CASE(alphabet, alphabet_types);
+
+TYPED_TEST(alphabet, default_value_constructor)
+{
+    TypeParam t1;
+    TypeParam t2{};
+}
+
+TYPED_TEST(alphabet, copy_constructor)
+{
+    TypeParam t1;
+    TypeParam t2{t1};
+    TypeParam t3(t1);
+    EXPECT_EQ(t1, t2);
+    EXPECT_EQ(t2, t3);
```

Comment:
> this makes no sense, because you can't be sure that anything was copied.
> You have to assign a value to t1, other than the default initialized.

The default initialization does not zero the object per se, the value is *indeterminate*.  I assumed that one can treat *indeterminate* as something like random, and that copying *indeterminate* values can be checked by this cascade. While this works (other than the fact that often the variables are indeed zero'ed by the compiler) by the standard it is UB so maybe we should change it.
It is, however, not an option to have the copy constructor tests use `assign_rank` and the assign_rank tests use the copy constructor. There must only be linear dependencies between the test cases.